gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/s3_signer.rb

```diff
@@ -163,7 +166,6 @@ class << self
           def new_hostname(context, region)
             endpoint_params = context[:endpoint_params].dup
             endpoint_params.region = region
-            endpoint_params.endpoint = nil
```

Comment:
@alextwoods do you think after this change we should remove the regional_endpoint check? because if we don't assign it nil it will be the value of metadata endpoint and will construct the host with the correct endpoint. I honestly didn't touch it becuase i couldn't found corresponding specs and got afraid to break something

Reply:
I was honestly on the fence about this - in the case you described, you are first getting a bucket redirect from an actual S3 bucket, then using a client with a custom endpoint, but using that same bucket key - in this case, you don't really want to assume anything about the bucket's region (since it may not be an actual S3 bucket).  I'd lean towards keeping the check for `regional_endpoint` as well as this fix.

Reply:
I think this line should be added back. Checking `regional_endpoint` config is correct. In `new_hostname`, we should assume it is a regional endpoint, so the endpoint provider should not be passed an endpoint.

Reply:
It might already default to nil here, but I think it makes sense to just set it to be nil to be sure, since we know we want this case.